server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/LocalAlignedGroupByExecutor.java

```diff
@@ -0,0 +1,388 @@
+      for (AggregateResult result : resultsOfOneMeasurement) {
+        if (!result.hasFinalResult()) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  public List<List<AggregateResult>> calcAlignedResult(long curStartTime, long curEndTime)
+      throws IOException, QueryProcessException {
+
+    // clear result cache
+    for (List<AggregateResult> resultsOfOneMeasurement : results) {
+      for (AggregateResult result : resultsOfOneMeasurement) {
+        result.reset();
+      }
+    }
+
+    timeRange.set(curStartTime, curEndTime - 1);
+    if (calcFromCacheData(curStartTime, curEndTime)) {
+      return results;
+    }
+
+    // read page data firstly
+    if (readAndCalcFromPage(curStartTime, curEndTime)) {
+      return results;
+    }
+
+    // read chunk data secondly
+    if (readAndCalcFromChunk(curStartTime, curEndTime)) {
+      return results;
+    }
+
+    // read from file
+    while (reader.hasNextFile()) {
+      // try to calc from fileMetaData
+      Statistics fileStatistics = reader.currentFileStatistics();
+      if (fileStatistics != null && !isEmpty(fileStatistics)) {
+        if (fileStatistics.getStartTime() >= curEndTime) {
+          return results;
+        }
+        if (reader.canUseCurrentFileStatistics()
+            && timeRange.contains(fileStatistics.getStartTime(), fileStatistics.getEndTime())) {
+          // calc from fileMetaData
+          while (reader.hasNextSubSeries()) {
+            Statistics currentFileStatistics = reader.currentFileStatistics();
+            calcFromStatistics(currentFileStatistics, results.get(reader.getCurIndex()));
+            reader.nextSeries();
+          }
+          reader.skipCurrentFile();
+          continue;
+        }
+      }
+      // read chunk
+      if (readAndCalcFromChunk(curStartTime, curEndTime)) {
+        return results;
+      }
+    }
+
+    return results;
+  }
+
+  private void calcFromStatistics(Statistics statistics, List<AggregateResult> aggregateResultList)
+      throws QueryProcessException {
+    // statistics may be null for aligned time series
+    if (statistics == null) {
+      return;
+    }
+    for (AggregateResult result : aggregateResultList) {
+      if (result.hasFinalResult()) {
+        continue;
+      }
+      result.updateResultFromStatistics(statistics);
+    }
+  }
+
+  private boolean readAndCalcFromChunk(long curStartTime, long curEndTime)
+      throws IOException, QueryProcessException {
+    while (reader.hasNextChunk()) {
+      // try to calc from chunkMetaData
+      Statistics chunkStatistics = reader.currentChunkStatistics();
+      if (chunkStatistics != null && !isEmpty(chunkStatistics)) {
+        if (chunkStatistics.getStartTime() >= curEndTime) {
+          if (ascending) {
+            return true;
+          } else {
+            reader.skipCurrentChunk();
+            continue;
+          }
+        }
+        if (reader.canUseCurrentChunkStatistics()
```

Comment:
change `canUseCurrentChunkStatistics` function to use time column in another pr.

Reply:
fix in #4451